README.MD
Pocket sized debug module for Phaser. Displays FPS graph, WEBGL drawcalls in any DOM (text) element. [Demo](https://samid737.github.io/phaser-plugin-pocketdebug/)

# Description

Displays FPS, MS graphs and drawcalls (WEBGL) in HTML DOM elements. Based upon 
[pocketplot](https://github.com/samid737/pocketplot), a simple plotting program.

A graph is text only and consists of scanlines:

```javascript
   _________*_*_____________*____ scanLine 0 (32 bit number)
   _______*_____*_________*______ scanLine 1
   _____*_________*__*_*_________ etc....
   _*_*__________________________ ....
   ______________________________ ..
```

# Requirements:

* Phaser v2 (tested in 2.6.2 and 2.8.4)
* Phaser.WEBGL mode for draw counts

# Usage

```javascript
pocketdebug =game.plugins.add(Phaser.Plugin.PocketDebug);

pocketdebug =game.plugins.add(Phaser.Plugin.PocketDebug,x,y,scale,advanced,bitmode);

//arguments : 
//  x,y             : position of panels relative to top-left corner of game
//  scale           : font-size and width of panel element multiplied by this scalevalue(default 1)
//  advanced        : adds an extra panel to measure state,stage,plugins,tweens,sound,input,physics,particles update time (in microseconds. 1 us=0.001ms).
//  bitmode         : do not reformat scanline, will output 0's and 1's, for retro feel and less overhead.

```
* UI:

* --    :   hide/show graphs.
* \+     :   zoom in Y range.
* \-     :   zoom out & range.
* \->    :   cycle trough graphs.
* \>>    :   increase refresh rate, resets after 10 clicks.

* click on A graph to change the color for better color contrast.

# Dev notes

the src folder contains a commented version. Ideas, contributions, optimizations, suggestions are very welcome. 
All changes are found in the [Changelog](https://github.com/samid737/phaser-plugin-pocketdebug/blob/master/CHANGELOG.md);
The source code for the wrapper function is from phaser-debug by englercj : https://github.com/englercj/phaser-debug.



CHANGELOG.md
# Change Log

## Unreleased

### Updates

### Bug Fixes

### Documentation

### Thanks

## Version 1.0.6 - 9th September 2017

### Updates

* Switch colors by tapping graph.
* Cycle trough graphs with the new panel layout.
* Added UI buttons to: toggle, resize, speedup, cycle.
* Added detail mode to measure state,stage,plugins,tweens,sound,input,physics,particles updates.

### Bug Fixes

### Documentation

* Usage: UI

### Thanks

@samid737

## Version 1.0.5 - 5th September 2017

### Updates

* added color change via hue shift
* added graph visibility toggling
* reverted to 1.02
* bitmode re-added, for supreme retro feel.

### Bug Fixes

### Documentation

### Thanks

@samid737

## Version 1.0.3 - 4th September 2017

### Updates

* removed FPS update 
* replaced string formatting (0 to _ and 1 to *) with custom font. up to 50% decrease in execution time
* removed bitmode

### Bug Fixes

### Documentation

### Thanks

@samid737

## Version 1.0.2 - 4th September 2017

### Updates

* bitmode added for retro feel (up to 50 % reduced execution time).

### Bug Fixes

### Documentation

### Thanks

@samid737


src/Graph.js
//a Graph is five scanlines concatenated to a single string
//
//   _________*_*_____________*____ scanLine 0
//   _______*_____*_________*______ scanLine 1
//   _____*_________*__*_*_________ scanLine 2
//   _*_*__________________________ scanLine 3
//   ______________________________ scanLine 4
//
//arguments :
//  game            : reference to current game 
//  maxY (in dev)   : the highest value of your input variable
//  refreshRate     : draw graph every refreshRate ms
//  bits            : retro mode, do not reformat 0's and 1's
//  label (in dev)  : the label of your input variable

var Graph = function (game,maxY,refreshRate,bits,label,content)
{
    this.game=game;this.refreshRate=this.initialRate=refreshRate;this.maxY=maxY;this.label=label;this.bits=bits;this.hide=false;this.content=content;
    //explained in Graph.draw(), the zeros are used to pad the final string result.
    this.scanBinary=this.startBinary=0x010000000;this.zeros= Array(30).join("0"); this.counter=0;
    // add the scanlines      
    this.line0=new Scanline(this,0);this.line1=new Scanline(this,1);
    this.line2=new Scanline(this,2);this.line3=new Scanline(this,3);
    this.line4=new Scanline(this,4);
    return this;
}

//graph drawing, if not hidden, concatenate output of Scanline.draw(), then returns the string

Graph.prototype.update =function(input)
{  
    //clock
    this.counter=(this.counter+1)%this.refreshRate;
    if(!this.hide&&this.counter==0)
    {
        //determine which scanline to draw on.
        this.input=input<1?input*1000:input;
        this.rownumber=~~((this.input-1)/(this.maxY/5));
        //scanbinary is a reference used for masking.if reached MSB, start over, else shift by one bit
        this.scanBinary=((this.scanBinary>>1))||this.startBinary;
        //concatenate and display resulting string
        this.result=this.line4.draw()+this.maxY+'\n'+this.line3.draw()+'\n'+this.line2.draw()+'\n'+this.line1.draw()+'\n'+this.line0.draw()+'\n';
        this.result+=this.input+" "+this.label+" DC:"+this.game.renderer.renderSession.drawCount;
        this.content.textContent=this.result;
    }
};



src/Panel.js

//a Panel is where all graphs are drawn into, the txtc and bgc are used in changeColor() with this.mask as mask bits.

var Panel =function(x,y,scale,components,dbg)
{
	this.game=dbg.game; this.graphs=[]; this.dbg=dbg;this.components=components;this.hide=false;this.nodes=[];this.index=0;
	this.width=340*scale;this.txtc=0xffffff;this.bgc=0xff00ff;this.mask=0x000016;this.fontSize=14;
	this.content=this.addUI(x,y,260,scale,0xff00ff,false,"",this,this.changeColor); 
	this.toggler=this.addUI(x,y,25,scale,0xffff00,true,"--",this,this.toggle);
	this.plus=this.addUI(x,y+17,25,scale,0x00ff00,true,"+",this,this.resize,1); 
	this.minus=this.addUI(x,y+34,25,scale,0xff0000,true,"-",this,this.resize,-1); 
	this.next=this.addUI(x,y+51,25,scale,0x0000ff,true,"->",this,this.change,1);  
	this.setspeed=this.addUI(x,y+65,25,scale,0xff0000,true,">>",this,this.speed,1);  
	return this;
};

//only update the current graph
Panel.prototype.update=function()
{
   this.current.update(this.components[this.current.label]);
};

Panel.prototype.destroy=function()
{
    this.nodes.forEach(function(element){this.game.canvas.parentNode.removeChild(element);});
};

//adds a data graph to the panel that can be cycled trough.

Panel.prototype.addGraph=function(maxY,refreshRate,bits,label)
{
    this.graph=new Graph(this.game,maxY,refreshRate,bits,label,this.content);
    this.graphs.push(this.graph);   
    this.current=this.graphs[0];
};

//add UI elements, if not a button don't label it. Bind click events to this panel.

Panel.prototype.addUI=function(x,y,width,scale,bgcolor,button,label,context,event,parameters)
{
    this.element=document.createElement("pre");
    this.element.setAttribute("style","position: absolute;left:"+(x+10)+"px;top:"+y+"px;width:"+(width*scale)+"px;text-align:center;color:white;font-weight:bold;font-size:"+11*scale+"px");
    this.game.canvas.parentNode.appendChild(this.element); 
    this.element.style["background-color"]=button?this.dbg.fastHexToRGB(bgcolor,0.3):this.dbg.fastHexToRGB(bgcolor,0.6); 
    this.element.textContent=button?label:0;
    this.element.onclick=event.bind(context,parameters);   
    this.nodes.push(this.element);   
    return this.element;
};

//Hue shift
Panel.prototype.changeColor=function()
{
    //shift left while not 180 degree hueshift
    this.mask=((this.mask<<1)&~0xff000000)||16;  
    //bitmasking
    this.bgc^=this.mask;
    this.txtc=~this.txtc;
    //console.log((("000000").substr( this.mask.toString(16).length)+ this.mask.toString(16)));//for debugging
    this.content.style["background-color"]=this.dbg.fastHexToRGB(this.bgc,0.6);
    this.content.style.color=this.dbg.fastHexToRGB(this.txtc,1);
};

//used to hide the graphs
Panel.prototype.toggle=function()
{  
    this.current.hide= this.content.hidden=!this.current.hide;
};

//resize the Y range of the graph (resize element in dev)

Panel.prototype.resize=function(direction)
{
    this.current.maxY=direction==1?(this.current.maxY*2):(this.current.maxY/2);
    //this.width+=40*direction;
    //this.fontSize+=direction;
    //this.content.style["width"]=this.width;
    //this.content.style["fontSize"]=this.fontSize;    
};

//switch graphs, current is the currently displayed graph

Panel.prototype.change=function()
{ 
    this.current.hide=true;
    this.index=(this.index+1)%(this.graphs.length);
    this.current=this.graphs[this.index];
    this.current.hide=false;
    //this.graphs.forEach(function(graph){console.log(graph.hide + " "+graph.label)});
};

//speed up the graph, increase redness proportionally to refreshRate

Panel.prototype.speed=function()
{
    this.current.refreshRate=Math.abs(((this.current.refreshRate-1)%this.current.initialRate))||this.current.initialRate;
    this.setspeed.style["background-color"]=this.dbg.fastHexToRGB(0xff0000*this.current.refreshRate|0xff0000,0.3+0.05*(10-this.current.refreshRate));
};

pocketdebug.js

var PocketDebug =this.Phaser.Plugin.PocketDebug = function (game, parent,x,y,scale)
{
    Phaser.Plugin.call(this, game, parent);
    this.name="Phaser Pocket Debug Plugin";
    this.panels=[];
    this.parameters={FPS:this.game.time.fps,MS:this.game.time.elapsedMS};
    this.components={state:0,stage: 0,plugins: 0,tweens: 0,sound: 0,input: 0,physics: 0,particles: 0,};
    this.timer = (window.performance ? window.performance : Date);

    this.fastHexToRGB=function(hex,alpha){
        return 'rgba('+(hex>>16)+","+((hex>>8)&(0x0000ff))+","+(hex&0x0000ff)+","+alpha+")";
    };
};

Phaser.Plugin.PocketDebug.prototype = Object.create(Phaser.Plugin.prototype);
Phaser.Plugin.PocketDebug.prototype.constructor = Phaser.Plugin.PocketDebug;

PocketDebug.prototype.init = function(x,y,scale,advanced,bits) 
{  
    this.x=x||0;this.y =y||0;this.scale=scale||1;this.advanced=advanced||false;
    this.game.time.advancedTiming = true;
    this.addPanel(this.x,0+this.y,this.scale,this.parameters,this);  
    this.panels[0].addGraph(60,10,bits,"FPS");
    this.panels[0].addGraph(20,10,bits,"MS");
    if(this.advanced){
        this.addPanel(this.x,100+this.y,this.scale,this.components,this);      
        for (var component in this.components) 
        {
            this.wrap(this.game, component, "update", component);
            this.panels[1].addGraph(1000,10,bits,component);  
        }
    }
};

PocketDebug.prototype.addPanel = function(x,y,scale,maxY,components,dbg)
{
    this.panel=new Panel(x,y,scale,maxY,components,dbg);
    this.panels.push(this.panel);  
};

PocketDebug.prototype.update=function()
{
    this.parameters.FPS=this.game.time.fps;
    this.parameters.MS=this.game.time.elapsedMS;
    this.panels[0].update();
    if(this.advanced){
        this.panels[1].update();
    }
};

PocketDebug.prototype.destroy = function()
{
    this.panels[0].destroy();
    this.panels[1].destroy();
    this.panels=null;      
    Phaser.Plugin.prototype.destroy.apply(this,arguments);        
};

PocketDebug.prototype.wrap = function (obj, component, method, timingStat) 
{
    obj[component][method] = (function(self, name, method, stat, fn)
    {    
        var t0, t1=0;
        return function () 
        {
        t0 = self.timer.now();
        fn.apply(this, arguments);
        t1 = self.timer.now();
        self.components[stat] = (t1 - t0).toFixed(2);//edit this for multi
        }
    })(this, component, method, timingStat, obj[component][method]);
};

var Panel =function(x,y,scale,components,dbg)
{
	this.game=dbg.game; this.graphs=[]; this.dbg=dbg;this.components=components;this.hide=false;this.nodes=[];this.index=0;
	this.width=340*scale;this.txtc=0xffffff;this.bgc=0xff00ff;this.mask=0x000016;this.fontSize=14;
	this.content=this.addUI(x,y,260,scale,0xff00ff,false,"",this,this.changeColor); 
	this.toggler=this.addUI(x,y,25,scale,0xffff00,true,"--",this,this.toggle);
	this.plus=this.addUI(x,y+17,25,scale,0x00ff00,true,"+",this,this.resize,1); 
	this.minus=this.addUI(x,y+34,25,scale,0xff0000,true,"-",this,this.resize,-1); 
	this.next=this.addUI(x,y+51,25,scale,0x0000ff,true,"->",this,this.change,1);  
	this.setspeed=this.addUI(x,y+65,25,scale,0xff0000,true,">>",this,this.speed,1);  
	return this;
};

Panel.prototype.update=function()
{
   this.current.update(this.components[this.current.label]);
};

Panel.prototype.destroy=function()
{
    this.nodes.forEach(function(element){this.game.canvas.parentNode.removeChild(element);});
};

Panel.prototype.addGraph=function(maxY,refreshRate,bits,label)
{
    this.graph=new Graph(this.game,maxY,refreshRate,bits,label,this.content);
    this.graphs.push(this.graph);   
    this.current=this.graphs[0];
};

Panel.prototype.addUI=function(x,y,width,scale,bgcolor,button,label,context,event,parameters)
{
    this.element=document.createElement("pre");
    this.element.setAttribute("style","position: absolute;left:"+(x+10)+"px;top:"+y+"px;width:"+(width*scale)+"px;text-align:center;color:white;font-weight:bold;font-size:"+11*scale+"px");
    this.game.canvas.parentNode.appendChild(this.element); 
    this.element.style["background-color"]=button?this.dbg.fastHexToRGB(bgcolor,0.3):this.dbg.fastHexToRGB(bgcolor,0.6); 
    this.element.textContent=button?label:0;
    this.element.onclick=event.bind(context,parameters);   
    this.nodes.push(this.element);   
    return this.element;
};
Panel.prototype.changeColor=function()
{
    this.mask=((this.mask<<1)&~0xff000000)||16;  
    this.bgc^=this.mask;
    this.txtc=~this.txtc;
    this.content.style["background-color"]=this.dbg.fastHexToRGB(this.bgc,0.6);
    this.content.style.color=this.dbg.fastHexToRGB(this.txtc,1);
};
Panel.prototype.toggle=function()
{  
    this.current.hide= this.content.hidden=!this.current.hide;
};

Panel.prototype.resize=function(direction)
{
    this.current.maxY=direction==1?(this.current.maxY*2):(this.current.maxY/2);
};

Panel.prototype.change=function()
{ 
    this.current.hide=true;
    this.index=(this.index+1)%(this.graphs.length);
    this.current=this.graphs[this.index];
    this.current.hide=false;
};

Panel.prototype.speed=function()
{
    this.current.refreshRate=Math.abs(((this.current.refreshRate-1)%this.current.initialRate))||this.current.initialRate;
    this.setspeed.style["background-color"]=this.dbg.fastHexToRGB(0xff0000*this.current.refreshRate|0xff0000,0.3+0.05*(10-this.current.refreshRate));
};

var Graph = function (game,maxY,refreshRate,bits,label,content)
{
    this.game=game;this.refreshRate=this.initialRate=refreshRate;this.maxY=maxY;this.label=label;this.bits=bits;this.hide=false;this.content=content;
    this.scanBinary=this.startBinary=0x010000000;this.zeros= Array(30).join("0"); this.counter=0;
    this.line0=new Scanline(this,0);this.line1=new Scanline(this,1);
    this.line2=new Scanline(this,2);this.line3=new Scanline(this,3);
    this.line4=new Scanline(this,4);
    return this;
}

Graph.prototype.update =function(input)
{  
    this.counter=(this.counter+1)%this.refreshRate;
    if(!this.hide&&this.counter==0)
    {
        this.input=input<1?input*1000:input;
        this.rownumber=~~((this.input-1)/(this.maxY/5));
        this.scanBinary=((this.scanBinary>>1))||this.startBinary;
        this.result=this.line4.draw()+this.maxY+'\n'+this.line3.draw()+'\n'+this.line2.draw()+'\n'+this.line1.draw()+'\n'+this.line0.draw()+'\n';
        this.result+=this.input+" "+this.label+" DC:"+this.game.renderer.renderSession.drawCount;
        this.content.textContent=this.result;
    }
};

var Scanline=function(gr,linenumber)
{
    this.gr=gr;this.linenumber=linenumber;this.binary=0;
    return  this;      
};

Scanline.prototype.draw=function()
{
    this.binary=this.gr.scanBinary==1?0:(this.gr.rownumber== this.linenumber?(this.gr.scanBinary^this.binary):(this.binary&~this.gr.scanBinary));//this.binary=this.gr.scanBinary; 
    this.n=this.binary.toString(2);
    this.n=this.gr.bits?(this.gr.zeros.substr( this.n.length)+ this.n):(this.gr.zeros.substr( this.n.length)+ this.n).replace(/0/g, "_").replace(/1/g, '*');
    return this.n;
};

src/Scanline.js
//a scanline is a 32 bit integer:  00000000000000000000000000000000

var Scanline=function(gr,linenumber)
{
    this.gr=gr;this.linenumber=linenumber;this.binary=0;
    return  this;      
};

//draw the scanLine  --> _________________________*______ 

Scanline.prototype.draw=function()
{
    //if reached MSB then reset scanline, else, if this fps is at my level, draw at my scanline, else silence me (mask the bits)
    this.binary=this.gr.scanBinary==1?0:(this.gr.rownumber== this.linenumber?(this.gr.scanBinary^this.binary):(this.binary&~this.gr.scanBinary));//this.binary=this.gr.scanBinary; 
    //this.binary=this.gr.scanBinary;//for debugging , displays default scanBinary
    this.n=this.binary.toString(2);
    //if in bitMode do not replace 0's and 1's
    this.n=this.gr.bits?(this.gr.zeros.substr( this.n.length)+ this.n):(this.gr.zeros.substr( this.n.length)+ this.n).replace(/0/g, "_").replace(/1/g, '*');
    return this.n;
};

src/PocketDebug.js
//Phaser Pocket Debug Plugin by samid737
//Pocket sized debug graph fit into a DOM text element.
//Displays FPS and MS graphs using strings.
var PocketDebug =this.Phaser.Plugin.PocketDebug = function (game, parent,x,y,scale)
{
    Phaser.Plugin.call(this, game, parent);
    this.name="Phaser Pocket Debug Plugin";
    this.panels=[];
    this.parameters={FPS:this.game.time.fps,MS:this.game.time.elapsedMS};
    this.components={state:0,stage: 0,plugins: 0,tweens: 0,sound: 0,input: 0,physics: 0,particles: 0,};
    this.timer = (window.performance ? window.performance : Date);

    this.fastHexToRGB=function(hex,alpha){
        return 'rgba('+(hex>>16)+","+((hex>>8)&(0x0000ff))+","+(hex&0x0000ff)+","+alpha+")";
    };
};

Phaser.Plugin.PocketDebug.prototype = Object.create(Phaser.Plugin.prototype);
Phaser.Plugin.PocketDebug.prototype.constructor = Phaser.Plugin.PocketDebug;

PocketDebug.prototype.init = function(x,y,scale,advanced,bits) 
{  
    this.x=x||0;this.y =y||0;this.scale=scale||1;this.advanced=advanced||false;
    this.game.time.advancedTiming = true;
    this.addPanel(this.x,0+this.y,this.scale,this.parameters,this);  
    this.panels[0].addGraph(60,10,bits,"FPS");
    this.panels[0].addGraph(20,10,bits,"MS");
    //wrap update functions
    if(this.advanced){
        this.addPanel(this.x,100+this.y,this.scale,this.components,this);      
        for (var component in this.components) 
        {
            this.wrap(this.game, component, "update", component);
            this.panels[1].addGraph(1000,10,bits,component);  
        }
    }
};

//add a graph to the DOM and update this graph using timer events
//two DOM elements are added, one contained with the actual graph, and another is a mask on top of the graph to cancel out zeros.

PocketDebug.prototype.addPanel = function(x,y,scale,maxY,components,dbg)
{
    this.panel=new Panel(x,y,scale,maxY,components,dbg);
    this.panels.push(this.panel);  
};

PocketDebug.prototype.update=function()
{
    this.parameters.FPS=this.game.time.fps;
    this.parameters.MS=this.game.time.elapsedMS;
    this.panels[0].update();
    if(this.advanced){
        this.panels[1].update();
    }
};

PocketDebug.prototype.destroy = function()
{
    this.panels[0].destroy();
    this.panels[1].destroy();
    this.panels=null;      
    Phaser.Plugin.prototype.destroy.apply(this,arguments);        
};

//timing wrapper, source from phaser-debug by englercj : https://github.com/englercj/phaser-debug

PocketDebug.prototype.wrap = function (obj, component, method, timingStat) 
{
    obj[component][method] = (function(self, name, method, stat, fn)
    {    
        var t0, t1=0;
        return function () 
        {
        t0 = self.timer.now();
        fn.apply(this, arguments);
        t1 = self.timer.now();
        self.components[stat] = (t1 - t0).toFixed(2);//edit this for multi
        }
    })(this, component, method, timingStat, obj[component][method]);
};



src/pocketdebugcolor.js
//Phaser Pocket Debug Plugin by samid737
//Pocket sized debug module that displays debug info in a DOM text element.
//Displays FPS and MS graph using bitwise and logical operators.
//https://github.com/samid737/phaser-pocket-debug

var PocketDebug =this.Phaser.Plugin.PocketDebug = function (game, parent) {
  Phaser.Plugin.call(this, game, parent);
  this.name="Phaser Pocket Debug";
  this.graphs={};
}

Phaser.Plugin.PocketDebug.prototype = Object.create(Phaser.Plugin.prototype);
Phaser.Plugin.PocketDebug.prototype.constructor = Phaser.Plugin.PocketDebug;

PocketDebug.prototype.init = function() {
  this.game.time.advancedTiming = true;
};
//add a graph to the DOM and update this graph using timer events
PocketDebug.prototype.add = function(x,y,scale,refreshRate,maxY,label,input) {
  this.gr=new Graph(game,x,y,scale,refreshRate,maxY,label);
  this.gr.node=document.createElement("pre"); 
  this.gr.node.setAttribute("style", "background-color: rgba(0, 0, 0, 0.4); position: absolute;left:"+x+"px;top:"+y+"px;width:"+(32*10*this.gr.scale)+"px;text-align:justify;font-size:"+14*this.gr.scale+"px");
  this.game.canvas.parentNode.appendChild(this.gr.node); 
  this.game.time.events.loop(refreshRate, this.gr.draw,this.gr);
  this.graphs[this.gr.label]=this.gr;
  return this.gr;
};
//destroy me
PocketDebug.prototype.destroy = function() {
  for(var graph in this.graphs){
    this.game.canvas.parentNode.removeChild(this.graphs[graph].node);
    this.graphs[graph]=null;    
  }  
  Phaser.Plugin.prototype.destroy.apply(this,arguments);        
};
//a Graph is five scanlines concatenated to a single string
//
//   _________*_*_____________*____ scanLine 0 all 32 bit numbers
//   _______*_____*_________*______ scanLine 1
//   _____*_________*__*_*_________ scanLine 2
//   _*_*__________________________ scanLine 3
//   ______________________________ scanLine 4
//
//arguments :
//  game            : reference to current game 
//  x,y             : position of graph relative to top-left corner of game
//  scale           : font-size and width of <pre> times this scalevalues(default 1)
//  refreshRate     : draw graph every refreshRate ms
//  maxY (in dev)   : the highest value of your input variable
//  label (in dev)  : the label of your input variable
//  input (in dev)  : your input variable
var Graph = function (game,x,y,scale,refreshRate,maxY,label){
  this.game=game,this.scale=scale,this.label=label,this.maxY=maxY;
  //explained in Graph.draw(), the zeros are used to pad the final string result.
  this.scanBinary=this.shiftCount=0;this.zeros= Array(37).join("0");
  // add the 5 scanlines      
  this.line0=new Scanline(this,0),this.line1=new Scanline(this,1),
  this.line2=new Scanline(this,2),this.line3=new Scanline(this,3),
  this.line4=new Scanline(this,4);
  return this;
}
//graph drawing, concatenates Scanline.draw() outputs, then displays the string
Graph.prototype.draw =function(){
  //which input
  this.input=this.label=="FPS"?this.game.time.fps:this.game.time.elapsedMS;
  //60 fps->row 0, 50 fps->row 1 , 40 fps->row 2.......
  this.rownumber=~~((this.input)/(this.maxY/5));
  //scanbinary is a reference number (zero lag binary).if reached MSB, reset to zero, else shift by one bit
  this.scanBinary=1>>this.scanBinary==0?0:1<<( this.shiftCount);
  //accumulator, used for scanBinary (see comment at 91); 
  this.shiftCount=1^this.shiftCount;
  //concatenate and display resulting string
  this.result=this.line4.draw()+this.line3.draw()+this.line2.draw()+this.line1.draw()+this.line0.draw();
  this.result+=this.input+" "+this.label+ " DC: "+this.game.renderer.renderSession.drawCount;
  this.node.textContent=this.result;
  //RGB---> 0 -> red , 255-> green
  this.node.style.color ='rgba('+(255-this.coloring())+', '+this.coloring()+',0,1.0)';
};
//graph coloring. 
Graph.prototype.coloring = function() {
  return ~~((this.game.time.fps/61)*255);
};
//a scanline is a 32 bit integer:  00000000000000000000000000000000
var Scanline=function(gr,linenumber){
  this.gr=gr,this.linenumber=linenumber,this.binary=0;
  return  this;      
} 
//draw the scanLine  --> _________________________*______ 
Scanline.prototype.draw=function(){
  //if reached MSB then reset scanline, else, if this fps is at my level, draw at my scanline, else silence me
  this.binary=1>> this.gr.scanBinary==0?0:(this.gr.rownumber== this.linenumber?(this.gr.scanBinary^this.binary):(this.binary&~this.gr.scanBinary));//this.binary=this.gr.scanBinary;//for debugging   
  this.n=this.binary.toString(2); 
  this.n=(this.gr.zeros.substr( this.n.length)+ this.n).replace(/0/g, "_").replace(/1/g, '*');
  return this.n+'\n';
}
